sky/data/storage.py

```diff
@@ -1871,6 +1902,495 @@ def _delete_gcs_bucket(self, bucket_name: str) -> bool:
+                    'and 63 (max) characters long.')
+
+            # Check for valid characters and start/end with a letter or number
+            pattern = r'^[a-z0-9][-a-z0-9]*[a-z0-9]$'
+            if not re.match(pattern, name):
+                _raise_no_traceback_name_error(
+                    f'Invalid store name: name {name} can consist only of '
+                    'lowercase letters, numbers, and hyphens (-). '
+                    'It must begin and end with a letter or number.')
+
+            # Check for two adjacent hyphens
+            if '--' in name:
+                _raise_no_traceback_name_error(
+                    f'Invalid store name: name {name} must not contain '
+                    'two adjacent hyphens.')
+
+        else:
+            _raise_no_traceback_name_error('Store name must be specified.')
+        return name
+
+    def initialize(self):
+        """Initializes the AZ Container object on the cloud.
+
+        Initialization involves fetching bucket if exists, or creating it if
+        it does not. Also, it checks for the existance of the storage account
+        and resource group if provided by the user. If not provided, those are
+        created with a default naming convention.
+
+        Raises:
+            StorageBucketCreateError: If bucket creation fails
+            StorageBucketGetError: If fetching existing bucket fails
+            StorageInitError: If general initialization fails.
+        """
+        self.storage_client = data_utils.create_az_client('storage')
+        self.resource_client = data_utils.create_az_client('resource')
+        # Using externally created storage
+        if isinstance(self.source, str) and self.source.startswith('az://'):
+            bucket_name, _, storage_account_name = data_utils.split_az_path(
+                self.source)
+            assert self.name == bucket_name
+            self.storage_account_name = storage_account_name
+            self.resource_group_name = data_utils.get_az_resource_group(
+                storage_account_name)
+        # Creating new resource group and storage account
+        else:
+            # If resource group or storage account names are not provided from
+            # config, then use default names.
+            self.storage_account_name = skypilot_config.get_nested(
+                ('azure', 'storage_account'),
+                f'sky{common_utils.get_user_hash()}')
+            self.resource_group_name = skypilot_config.get_nested(
+                ('azure', 'resource_group'),
+                f'sky{common_utils.get_user_hash()}')
+            # If the resource group name and storage account name used to run
+            # create functions below already exist under user's
+            # subscription id, both will silently move on.
+            self.resource_client.resource_groups.create_or_update(
+                self.resource_group_name, {'location': self.region})
```

Comment:
After fixing the permissions, the next `sky spot launch` failed with this error:

```
I 05-01 13:21:51 controller_utils.py:602] Uploading sources to cloud storage. See: sky storage ls
Traceback (most recent call last):
  File "/Users/romilb/tools/anaconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 354, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 805, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 354, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 805, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 3292, in spot_launch
    spot_lib.launch(dag,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 375, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/spot/core.py", line 77, in launch
    controller_utils.maybe_translate_local_file_mounts_and_sync_up(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/controller_utils.py", line 604, in maybe_translate_local_file_mounts_and_sync_up
    task.sync_storage_mounts()
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/task.py", line 930, in sync_storage_mounts
    storage.add_store(store_type, store_region)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 855, in add_store
    store = store_cls(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1924, in __init__
    super().__init__(name, source, region, is_sky_managed,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 242, in __init__
    self.initialize()
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2056, in initialize
    self.resource_client.resource_groups.create_or_update(
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/mgmt/resource/resources/v2021_04_01/operations/_operations.py", line 9810, in create_or_update
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceExistsError: (InvalidResourceGroupLocation) Invalid resource group location 'westus3'. The Resource group already exists in location 'eastus'.
Code: InvalidResourceGroupLocation
Message: Invalid resource group location 'westus3'. The Resource group already exists in location 'eastus'.
```

We probably need to fetch and reuse the location of an existing group? Also need to handle the case when the requested region is different than the existing resource group's region.

Reply:
This error was caused as I allowed the creation logic for resource group and storage account to run even though they already existed. That's why in the error you encountered, it was trying to create another resource group using the same name but in different region following the spot instance.

I added a check to see if the resource group and storage account of a given name already exists under the user's account to fix the issue. Now, as it would already be existing, it will not try to create the already existing resource group.

Reply:
Regards to handling the case when the requested region is different than the existing resource group's region, I updated so that the default resource group is created per region at a607874